File: src/game/systems/damage.lua
local Aspect = require 'ecs.aspect'
local System = require 'ecs.system'
local Damage = require 'game.components.damage'
local Fixture = require 'game.components.fixture'
local Health = require 'game.components.health'

local DamageSystem = System:new('damage', Aspect.always())

function DamageSystem:beginContact(a, b, contact)
  a = a:getUserData().entity
  b = b:getUserData().entity
  local health
  local damage

  if a:has(Damage) and b:has(Health) then
    damage = a:as(Damage)
    health = b:as(Health)
  elseif a:has(Health) and b:has(Damage) then
    health = a:as(Health)
    damage = b:as(Damage)
  end

  if health and damage then
    health.hitpoints = health.hitpoints - damage.hitpoints
  end
end

return DamageSystem

File: client2/src/game/systems/Damage.ts
import { System } from '../../ecs/System'
import { SystemFlag } from '../flags'
import { NeverAspect } from '../../ecs/Aspect'
import { check, hasComponent, isNotBodyType } from '../utils/collision'
import { Damage } from '../components/Damage'
import { Health } from '../components/Health'

export class DamageSystem extends System {
  static _id = 'Damage'
  _id = DamageSystem._id

  static _flag = SystemFlag.Damage
  _flag = DamageSystem._flag

  static _aspect = new NeverAspect()
  _aspect = DamageSystem._aspect

  beginContact = (a: Fixture, b: Fixture, contact: Contact) => {
    const result = check(a, b, [hasComponent(Damage), hasComponent(Health)])

    if (!result) {
      return
    }

    const damage = result[0].as(Damage)
    const health = result[0].as(Health)
    if (damage && health) {
      health.hitpoints = health.armor - damage.hitpoints
    }
  }
}
